collections-contrib/src/main/scala/strawman/collection/NonEmpty.scala

```diff
@@ -0,0 +1,90 @@
+package strawman
+package collection
+
+import collection.immutable.List
+
+/**
+  * A wrapper around a collection that is guaranteed to have at least one element.
+  *
+  * A `NonEmpty` instance is implicitly enriched with the same operations as the
+  * wrapped collection and also has its own operations that preserve non-emptiness.
+  *
+  * @param coll Wrapped collection
+  * @tparam A Type of elements (e.g. `Int`, `String`)
+  * @tparam C Type of collection (e.g. `List[Int]`, `Map[Int, Boolean]`)
+  */
+final class NonEmpty[+A, C <: Iterable[A]] private (val coll: C)
+  extends AnyVal {
+
+  def map[B, C2 <: Iterable[B]](f: A => B)(implicit bf: BuildFrom[C, B, C2]): NonEmpty[B, C2] =
+    new NonEmpty[B, C2](bf.fromSpecificIterable(coll)(coll.toIterable.map(f)))
```

Comment:
Why does this not just defer directly to the `map` method on `C`?

Reply:
(Same comment  about "why not directly" for everything else.)

Reply:
Indeed we could omit the `.toIterable` call. That’s a left over of a previous iteration…

Reply:
The `bf.fromSpecificIterable` call is needed because it allows us to compute the resulting collection type `C2` (otherwise we would get an `Iterable[B]`).